### ISSUE TITLE
[CLOSED] More Filter

### ISSUE THREAD

**Issue by llmhyy** [llmhyy/evosuite-plus-plus#24]
<a href="https://github.com/llmhyy"><img src="https://avatars1.githubusercontent.com/u/2909461?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [llmhyy](https://github.com/llmhyy)**
_Friday Feb 08, 2019 at 17:02 GMT_
_Originally opened as https://github.com/llmhyy/evosuite_progress/issues/24_

----

Hi LyLy,

I checked a few methods. I feel that those methods cannot show the advantage of our approach. Let me give two examples, 
Case I: there are a great number of equals() methods to compare object reference.
Case II: there are a lot of Iterator.hasNext() methods by calling "for(Obj obj: list)" syntax.
The above two cases are out of the boundary of our this research work. We should ignore those cases.

Thus, I suggest that we need to filter methods with the following rules (probably, we need to do it as soon as possible):
1. methods suffering from interprocedural flag problem (we already have that)
2. for the called method (with boolean return type), we need to check the source of its ireturn instruction.
More specifically, the method with boolean return type must end with an ireturn instruction. We need to further check the source of the ireturn instruction. Based on my observation, the scenarios can be
(1) iconst_0/1, i.e., return true/false in source code
(2) getfield, i.e., return obj.booleanVariable in source code
(3) invokevirtualmethod, i.e., return isOK() in source code.
We keep methods conforming to (1) and throw away those conforming to (2). While for (3), we should recursively analyze whether the invoked method (e.g., isOk()) conforming to (1) or (2). If (1), we keep it, otherwise, we throw it away. By using Evosuite API, you will find that the CFG of some called method is missing, in such case, if there is no iconst_0/1 found, we classify it to (2).

You may also kindly help consider record how many interprocedural flag method belongs to (1) and (2).

If there are very few methods belonging to (1), we need to find other Java projects for our work. The potential ones are those used for the SVM paper. 
https://github.com/sunjun-group/test-projects:
colt, jblas, jscience, and math.
My suggestion is that, after you finish the method filter, you can apply the code for all methods in SF-100 and the above four projects.

In summary, I list the TODO as follows:
1. write rules to filter methods
2. apply them on both SF-100 dataset and four math related Java projects

You may consider letting me run a part of them, as the server in SUTD will be used by only you and I from tomorrow night.

Many thanks for your kind help! Let me treat you more when you are back, and nicer when we get through the deadline!
